frontend/src/components/worksheets/NewUpload/NewUpload.jsx

```diff
@@ -41,54 +43,180 @@ class NewUpload extends React.Component<{
+                                if (index == files.length - 1) {
+                                    const moveIndex = true;
+                                    const param = { moveIndex };
+                                    this.props.reloadWorksheet(undefined, undefined, param);
+                                    this.props.onUploadFinish();
+                                }
+                            },
+                            error: (jqHXR, status, error) => {
+                                this.clearProgress();
+                                alert(
+                                    createAlertText(
+                                        reader.url,
+                                        jqHXR.responseText,
+                                        'refresh and try again.',
+                                    ),
+                                );
+                                this.props.onUploadFinish();
+                            },
+                        });
+                    };
+                    reader.readAsArrayBuffer(file);
+                },
+                error: (jqHXR, status, error) => {
+                    this.clearProgress();
+                    alert(createAlertText(url, jqHXR.responseText));
+                },
+            });
+        }
     }
 
-    uploadFile = (file) => {
-        if (!file) {
+    uploadFolder = (files) => {
+        if (!files) {
+            return;
+        }
+        if (files.length > 1000) {
+            alert('There are too many files in the folder. Please zip your folder and upload again.');
             return;
         }
+
         const { worksheetUUID, after_sort_key } = this.props;
         const { name, description } = this.state;
-        const createBundleData = getDefaultBundleMetadata(name || file.name, description);
+        const folderNamePos = files[0].webkitRelativePath.indexOf("/");
+        let folderName = "";
+        if (folderNamePos != -1) {
+            folderName = files[0].webkitRelativePath.slice(0, folderNamePos)
+        }
+
+        const createBundleData = getDefaultBundleMetadata(name || folderName + ".zip", description);
         this.setState({
             uploading: true,
         });
         let url = `/rest/bundles?worksheet=${ worksheetUUID }`;
         if (after_sort_key) {
             url += `&after_sort_key=${ after_sort_key }`;
         }
-        $.ajax({
-            url,
-            data: JSON.stringify(createBundleData),
-            contentType: 'application/json',
-            type: 'POST',
-            success: (data, status, jqXHR) => {
-                var bundleUuid = data.data[0].id;
-                var reader = new FileReader();
-                reader.onload = () => {
-                    var arrayBuffer = reader.result,
-                        bytesArray = new Uint8Array(arrayBuffer);
-                    var url =
+
+        let zip = new JSZip();
+        [...files].map(file => {

```

Comment:
Can you just use files instead of ...files?

Reply:
File API treats files as an object, so we need to flatten it.